### DISCUSSION TITLE
Scheduling Rewrite

### DISCUSSION THREAD

**Discussion by DamienCassou** [borgbase/vorta#966] [Ideas]
Vorta saves my computer to 2 different locations: on a LAN backup server and on borgbase. My wife's laptop will soon have the same setup.

Question: How to best configure Vorta's scheduling so only one backup run at a time (so the LAN always stays available for other things)?

I thought about a scheme like this:

- even hour + 10 minutes : laptop1-backup1
- even hour + 40 minutes : laptop2-backup1
- odd hour + 10 minutes : laptop1-backup2
- odd hour + 40 minutes : laptop2-backup2

Beyond spreading the backups to avoid network flooding, this schemes increases the time between 2 backups of the same laptop: this way, Vorta won't complain that another backup is in progress when trying to start a new one.

Unfortunately, I can't express this scheme in Vorta. I can say "every 2 hours" but I don't know if it will be even or odd hours.

**Comment by m3nu** [comment]
I'm currently in the process of rewriting the scheduler, but it's not done yet.

https://github.com/borgbase/vorta/pull/908

Currently we use APScheduler, which may drop support for Qt in the next version. It's also overkill to have another dependency, when Qt already provides the QTimer class to do something similar (but lower level).

The new implementation will take the last backup time into account and offer to catch up with failed backups.

Queueing backups is an interesting aspect I didnt' consider yet.

If coding wouldn't matter, how would you present these options in the UI? To give you the options you need, but also keeping it simple?

**Reply by DamienCassou** [reply]
I think the current UI is great: it will be enough for most users I guess.

But then some users (like me) want something more advanced and I see several options:

1. tell them to use their own scheduler (e.g., a systemd timer) and to call vorta from the command line (there could be a "help" button in the UI which gives examples or redirects to a web page with examples);
2. make vorta engine support advanced schedules and ask the user to modify a vorta's configuration file (there could be a "help" button in the UI which explains what to do);
3. add an "advanced schedule" button that displays and advanced UI to setup the schedule. This UI will depend on the capabilities of QTimer which I'm not familiar with. If QTimer is close to crontab or systemd, maybe a text input field with a link to documentation and syntax validation would be enough;
4. Make it possible to setup several "Backup daily at:" times. This way, instead of expressing "even hour + 10 minutes", I would add these times: "08:10", "10:10", "12:10", "14:10", "16:10", "18:10", "20:10", "22:10";
5. Add a "Backup every XXX minutes starting at HH:MM" option where XXX, HH and MM are 3 number input fields. This way, I can have XXX=120, HH=8 and MM=10 to implement the schedule of the above option (even hours from 8:10 to 22:10). I think this option is strictly more powerful that the "Backup every XXX hours at YY minutes past the hour" that you have already. Maybe it could replace it.

**Reply by m3nu** [reply]
We probably don't have the resources to program an extra "Advanced Scheduling" setting. Are you able to contribute a pull request for such a feature?

Regarding external schedulers: You can already trigger a backup run via the command line, if Vorta is already running. Just run `vorta -h` to view the precise arguments. Something like `vorta --profile xxx`

**Reply by DamienCassou** [reply]
> Are you able to contribute a pull request for such a feature?

That won't be possible, I already have too much on my plate. Are you saying that none of the 5 options above satisfy you? 

> You can already trigger a backup run via the command line

Yes, I was aware of that. Thank you.

**Reply by m3nu** [reply]
We already support option 1.

For the other options: would be happy to implement them, but will focus on getting the basic right first.
